test/TokenRequestTest.js

```diff
@@ -1,51 +1,304 @@
+    beforeEach(async () => {
+      forwarderMockBase = await ForwarderMock.new()
+      const newForwarderMockReceipt = await daoDeployment.kernel.newAppInstance(
+        '0x9876',
+        forwarderMockBase.address,
+        '0x',
+        false,
+        { from: rootAccount }
+      )
+      forwarderMock = await ForwarderMock.at(deployedContract(newForwarderMockReceipt))
+
+      await forwarderMock.initialize()
+
+      await daoDeployment.acl.createPermission(tokenRequest.address, tokenManager.address, MINT_ROLE, rootAccount)
+      await daoDeployment.acl.createPermission(
+        forwarderMock.address,
+        tokenRequest.address,
+        FINALISE_TOKEN_REQUEST_ROLE,
+        rootAccount
+      )
+    })
+
+    it('finalise token request (ERC20)', async () => {
+      const expectedUserMiniMeBalance = 300
+      const expectedVaultBalance = 100
+
+      await mockErc20.approve(tokenRequest.address, ROOT_TOKEN_AMOUNT, {
+        from: rootAccount,
+      })
+      await tokenRequest.createTokenRequest(mockErc20.address, expectedVaultBalance, expectedUserMiniMeBalance, {
+        from: rootAccount,
+      })
+
+      const action = {
+        to: tokenRequest.address,
+        calldata: tokenRequest.contract.methods.finaliseTokenRequest(0).encodeABI(),
+      }
+      script = encodeCallScript([action])
+      await forwarderMock.forward(script, { from: rootAccount })
+
+      const actualUserMiniMeBalance = await tokenManager.spendableBalanceOf(rootAccount)
+      const actualVaultBalance = await vault.balance(mockErc20.address)
+
+      assert.equal(actualUserMiniMeBalance, expectedUserMiniMeBalance)
+      assert.equal(actualVaultBalance, expectedVaultBalance)
+    })
+
+    it('finalise token request (ETH)', async () => {
+      const expectedUserMiniMeBalance = 300
+      const expectedVaultBalance = 200
+
+      await tokenRequest.createTokenRequest(ZERO_ADDRESS, expectedVaultBalance, expectedUserMiniMeBalance, {
+        from: rootAccount,
+        value: expectedVaultBalance,
+      })
+
+      const action = {
+        to: tokenRequest.address,
+        calldata: tokenRequest.contract.methods.finaliseTokenRequest(0).encodeABI(),
+      }
+      script = encodeCallScript([action])
+      await forwarderMock.forward(script, { from: rootAccount })
+
+      const actualUserMiniMeBalance = await tokenManager.spendableBalanceOf(rootAccount)
+      const actualVaultBalance = await vault.balance(ZERO_ADDRESS)
+
+      assert.equal(actualUserMiniMeBalance, expectedUserMiniMeBalance)
+      assert.equal(actualVaultBalance, expectedVaultBalance)
+    })
+  })
+
+  describe('refundTokenRequest(uint256 _tokenRequestId) ', () => {
+    it('refound token (ERC20)', async () => {
+      const expectedRefoundAmount = 100
+
+      await mockErc20.approve(tokenRequest.address, expectedRefoundAmount, {
+        from: rootAccount,
+      })
+      await tokenRequest.createTokenRequest(mockErc20.address, expectedRefoundAmount, 1, {
```

Comment:
approve() and createTokenRequest() could be done in a beforeEach hook

Reply:
the create token request is different for each test